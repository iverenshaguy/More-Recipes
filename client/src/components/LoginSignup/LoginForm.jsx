import React from 'react';
import { TabPane, Form, Label, Input, FormGroup, Button } from 'reactstrap';

const LoginForm = () => (
  <TabPane tabId="1" id="signin-form-div" role="tabpanel">
    <h4 className="text-center">Sign In to Your Account</h4>
    <hr />
    <Form id="signin-form" className="mt-4 mb-3 px-5">
      <FormGroup>
        <Label for="login-email" className="col-form-label">
          Username/Email Address
        </Label>
        <Input type="email" id="login-email" placeholder="janesmith" />
      </FormGroup>
      <FormGroup>
        <Label for="login-password" className="col-form-label">
          Password
        </Label>
        <Input
          type="password"
          className="mb-4"
          id="login-password"
          placeholder={'\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022'}
        />
      </FormGroup>
      <FormGroup check className="mb-0">
        <Label check for="remember-me" className="col-form-label col-form-label-lg">
          <Input type="checkbox" className="form-check-input" id="remember-me" /> Remember Me
        </Label>
      </FormGroup>
      <Button className="btn-block mt-0">SIGN IN</Button>
    </Form>
  </TabPane>
);

export default LoginForm;
