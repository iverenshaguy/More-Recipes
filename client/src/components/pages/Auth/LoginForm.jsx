import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Form, Label, FormGroup, Button } from 'reactstrap';
import { RenderInput } from '../../shared/FormComponents';
import { asyncValidate, syncValidate } from '../../../helpers/validations';
import { NormalAlert } from '../../shared/Alert';
import setCurrentLocation from '../../../actions/location';
import { login, clearAuthError } from '../../../actions/auth';

const placeholder = '\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022';

/**
 * @exports
 * @class LoginForm
 * @extends Component
 * @returns {component} LoginForm
 */
class LoginForm extends Component {
  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    authenticating: PropTypes.bool.isRequired,
    submitError: PropTypes.string,
  };

  static defaultProps = {
    submitError: null
  }

  /**
   * @constructor
   * @memberof LoginForm
   * @returns {nothing} Return nothing
   */
  constructor() {
    super();

    this.state = {
      values: {
        email: '',
        password: ''
      },
      touched: { email: false, password: false },
      error: { email: null, password: null },
      pristine: true,
      formValid: false,
      asyncValidating: false
    };

    this.handleBlur = this.handleBlur.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleFocus = this.handleFocus.bind(this);
    this.validateField = this.validateField.bind(this);
    this.validateForm = this.validateForm.bind(this);
  }

  /**
   * @memberof LoginForm
   * @returns {nothing} Returns nothing
   */
  componentWillMount() {
    this.props.dispatch(setCurrentLocation('auth'));
    this.props.dispatch(clearAuthError());
  }

  /**
   * @memberof LoginForm
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleChange(event) {
    const { target } = event;
    const { value, name } = target;

    this.setState({
      values: { ...this.state.values, [name]: value },
      touched: { ...this.state.touched, [name]: true },
      pristine: false,
    }, () => { this.validateField(name, value); });
  }

  /**
   * @memberof LoginForm
   * @returns {nothing} Returns nothing
   */
  handleFocus() {
    this.props.dispatch(clearAuthError());
  }

  /**
   * @memberof LoginForm
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleBlur(event) {
    const { target } = event;
    const { value, name } = target;

    this.setState({
      touched: { ...this.state.touched, [name]: true },
    }, () => { this.validateField(name); });

    setTimeout(() => {
      if (name === 'email' && !this.state.error.email) {
        this.setState({
          asyncValidating: true
        }, () => {
          asyncValidate('login')(name, value)
            .then(() => {
              this.setState({
                asyncValidating: false
              });
            })
            .catch((error) => {
              this.setState({
                error: { ...this.state.error, email: error.email },
                asyncValidating: false
              });
            });
        });
      }
    }, 500);
  }

  /**
   * @memberof LoginForm
   * @param {string} field
   * @param {string} values
   * @returns {nothing} Returns nothing
   */
  validateField(field) {
    const error = syncValidate('login')(field, this.state.values);

    if (error) {
      this.setState({
        error: {
          ...this.state.error,
          [field]: error
        }
      }, this.validateForm);
    } else {
      this.setState({
        error: {
          ...this.state.error,
          [field]: null
        }
      }, this.validateForm);
    }
  }

  /**
   * @memberof LoginForm
   * @returns {nothing} Returns nothing
   */
  validateForm() {
    const formErrorArrayLength =
      Object.values(this.state.error).filter(value => value !== null).length;
    const formTouchedArrayLength =
      Object.values(this.state.touched).filter(value => value === true).length;
    const formFieldCount = Object.keys(this.state.values).length;

    if (formErrorArrayLength) {
      this.setState({
        formValid: false
      });
    } else if (!formErrorArrayLength && formTouchedArrayLength === formFieldCount) {
      this.setState({
        formValid: true
      });
    }
  }

  /**
   * @memberof LoginForm
   * @param {object} event
   * @returns {nothing} Returns nothing
   */
  handleSubmit(event) {
    event.preventDefault();
    this.props.dispatch(login(this.state.values));
  }

  /**
   * @memberof LoginForm
   * @returns {component} LoginForm
   */
  render() {
    return (
      <div id="signin-form-div">
        <h4 className="text-center">Sign In to Your Account</h4>
        <hr />
        <Form id="signin-form" className="mt-4 mb-3 px-5" onSubmit={this.handleSubmit}>
          {this.props.submitError && (
            <NormalAlert color="danger">
              <p className="text-center mb-0">{this.props.submitError}</p>
            </NormalAlert>
          )}
          <FormGroup>
            <Label for="login-email" className="col-form-label">
              Email Address
            </Label>
            <RenderInput
              type="email"
              name="email"
              id="login-email"
              value={this.state.values.email}
              placeholder="janesmith@me.com"
              handleChange={this.handleChange}
              handleBlur={this.handleBlur}
              handleFocus={this.handleFocus}
              meta={{
                touched: this.state.touched.email,
                error: this.state.error.email,
                asyncValidating: this.state.asyncValidating
              }}
            />
          </FormGroup>
          <FormGroup>
            <Label for="login-password" className="col-form-label">
              Password
            </Label>
            <RenderInput
              type="password"
              name="password"
              className="mb-4"
              id="login-password"
              value={this.state.values.password}
              placeholder={placeholder}
              handleChange={this.handleChange}
              handleBlur={this.handleBlur}
              handleFocus={this.handleFocus}
              meta={{
                touched: this.state.touched.password,
                error: this.state.error.password,
              }}
            />
          </FormGroup>
          <Button className="btn-block mt-0" disabled={!this.state.formValid || this.state.pristine || this.props.authenticating}>
            SIGN IN
          </Button>
        </Form>
        <p className="text-center">Dont have an account, <Link to="\signup" target="_self">Sign up here</Link>.</p>
      </div>
    );
  }
}


export { LoginForm as LoginComponent };

export default connect()(LoginForm);
