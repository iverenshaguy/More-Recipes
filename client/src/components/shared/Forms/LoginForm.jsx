import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { Form, Label, FormGroup, Button } from 'reactstrap';
import { RenderInput } from '../FormComponents';
import { NormalAlert } from '../Alert';

const placeholder = '\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022\u2022';

/**
 * @exports
 * @param {object} props
 * @returns {component} LoginForm
 */
const LoginForm = ({ state, handlers }) => (
  <div id="signin-form-div" >
    <h4 className="text-center">Sign In to Your Account</h4>
    <hr />
    <Form id="signin-form" className="mt-4 mb-3 px-5" onSubmit={handlers.handleSubmit}>
      {state.submitError && (
        <NormalAlert color="danger">
          <p className="text-center mb-0">{state.submitError}</p>
        </NormalAlert>
      )}
      <FormGroup>
        <Label for="login-email" className="col-form-label">
          Email Address
        </Label>
        <RenderInput
          type="email"
          name="email"
          id="login-email"
          value={state.values.email}
          placeholder="janesmith@me.com"
          handleChange={handlers.handleChange}
          handleBlur={handlers.handleBlur}
          handleFocus={handlers.handleFocus}
          meta={{
            touched: state.touched.email,
            error: state.error.email,
            asyncValidating: state.asyncValidating
          }}
        />
      </FormGroup>
      <FormGroup>
        <Label for="login-password" className="col-form-label">
          Password
        </Label>
        <RenderInput
          type="password"
          name="password"
          className="mb-4"
          id="login-password"
          value={state.values.password}
          placeholder={placeholder}
          handleChange={handlers.handleChange}
          handleBlur={handlers.handleBlur}
          handleFocus={handlers.handleFocus}
          meta={{
            touched: state.touched.password,
            error: state.error.password,
          }}
        />
      </FormGroup>
      <Button className="btn-block mt-0" disabled={!state.formValid || state.pristine || state.submitting}>
        SIGN IN
      </Button>
    </Form>
    <p className="text-center">Dont have an account, <Link to="\signup" target="_self">Sign up here</Link>.</p>
  </div>
);

LoginForm.propTypes = {
  state: PropTypes.shape({
    values: PropTypes.shape({
      email: PropTypes.string,
      password: PropTypes.string
    }),
    touched: PropTypes.shape({
      email: PropTypes.bool,
      password: PropTypes.bool
    }),
    error: PropTypes.shape({
      email: PropTypes.string,
      password: PropTypes.string
    }),
    pristine: PropTypes.bool,
    formValid: PropTypes.bool,
    asyncValidating: PropTypes.bool,
    submitting: PropTypes.bool,
    submitError: PropTypes.string
  }).isRequired,
  handlers: PropTypes.shape({
    handleBlur: PropTypes.func,
    handleChange: PropTypes.func,
    handleFocus: PropTypes.func,
    handleSubmit: PropTypes.func,
  }).isRequired
};

export default LoginForm;
