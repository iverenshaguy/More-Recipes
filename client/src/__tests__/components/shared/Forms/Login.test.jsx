import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import LoginForm from '../../../../components/shared/Forms/LoginForm';

const dispatchMock = jest.fn();

const setup = () => {
  const props = {
    state: {
      values: {
        email: '',
        password: ''
      },
      touched: { email: false, password: false },
      error: { email: null, password: null },
      pristine: true,
      formValid: false,
      asyncValidating: false,
      submitting: false,
      submitError: null
    },
    handlers: {
      handleBlur: jest.fn(),
      handleChange: jest.fn(),
      handleFocus: jest.fn(),
      handleSubmit: jest.fn()
    }

  };

  const mountRoot = mount( //eslint-disable-line
    <MemoryRouter>
      <LoginForm {...props} dispatch={dispatchMock} />
    </MemoryRouter>);

  return { props, mountRoot };
};

describe('LoginForm', () => {
  afterAll(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { props } = setup();
    const shallowComponent = shallow(<LoginForm {...props} />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
  });

  it('disables submit button when form is clean', () => {
    const { props } = setup();
    const shallowComponent = shallow(<LoginForm {...props} />);

    expect(shallowComponent.find('Button[disabled=true]')).toBeTruthy();
  });

  it('disables submit button when submitting form', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      state: { ...props.state, submitting: true }
    };
    const shallowComponent = shallow(<LoginForm {...newProps} />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
    expect(shallowComponent.find('Button[disabled=true]')).toBeTruthy();
  });

  it('shows error alert and disables submit button when there\'s a submit error', () => {
    const { props } = setup();
    const newProps = {
      ...props,
      state: { ...props.state, submitError: 'Username/Password do not match' }
    };

    const shallowComponent = shallow(<LoginForm {...newProps} />);

    expect(toJson(shallowComponent)).toMatchSnapshot();
    expect(shallowComponent.find('NormalAlert')).toBeTruthy();
    expect(shallowComponent.find('Button[disabled=true]')).toBeTruthy();
  });
});
