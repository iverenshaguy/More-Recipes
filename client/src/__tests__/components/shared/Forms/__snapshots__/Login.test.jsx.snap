// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm disables submit button when submitting form 1`] = `
<div
  id="signin-form-div"
>
  <h4
    className="text-center"
  >
    Sign In to Your Account
  </h4>
  <hr />
  <Form
    className="mt-4 mb-3 px-5"
    id="signin-form"
    onSubmit={[Function]}
    tag="form"
  >
    <FormGroup
      tag="div"
    >
      <Label
        className="col-form-label"
        for="login-email"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Email Address
      </Label>
      <RenderInput
        handleBlur={[Function]}
        handleChange={[Function]}
        handleFocus={[Function]}
        id="login-email"
        meta={
          Object {
            "asyncValidating": false,
            "error": null,
            "touched": false,
          }
        }
        name="email"
        placeholder="janesmith@me.com"
        type="email"
        value=""
      />
    </FormGroup>
    <FormGroup
      tag="div"
    >
      <Label
        className="col-form-label"
        for="login-password"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Password
      </Label>
      <RenderInput
        className="mb-4"
        handleBlur={[Function]}
        handleChange={[Function]}
        handleFocus={[Function]}
        id="login-password"
        meta={
          Object {
            "error": null,
            "touched": false,
          }
        }
        name="password"
        placeholder="••••••••••"
        type="password"
        value=""
      />
    </FormGroup>
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN IN
    </Button>
  </Form>
  <p
    className="text-center"
  >
    Dont have an account, 
    <Link
      replace={false}
      target="_self"
      to="\\\\signup"
    >
      Sign up here
    </Link>
    .
  </p>
</div>
`;

exports[`LoginForm renders correctly 1`] = `
<div
  id="signin-form-div"
>
  <h4
    className="text-center"
  >
    Sign In to Your Account
  </h4>
  <hr />
  <Form
    className="mt-4 mb-3 px-5"
    id="signin-form"
    onSubmit={[Function]}
    tag="form"
  >
    <FormGroup
      tag="div"
    >
      <Label
        className="col-form-label"
        for="login-email"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Email Address
      </Label>
      <RenderInput
        handleBlur={[Function]}
        handleChange={[Function]}
        handleFocus={[Function]}
        id="login-email"
        meta={
          Object {
            "asyncValidating": false,
            "error": null,
            "touched": false,
          }
        }
        name="email"
        placeholder="janesmith@me.com"
        type="email"
        value=""
      />
    </FormGroup>
    <FormGroup
      tag="div"
    >
      <Label
        className="col-form-label"
        for="login-password"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Password
      </Label>
      <RenderInput
        className="mb-4"
        handleBlur={[Function]}
        handleChange={[Function]}
        handleFocus={[Function]}
        id="login-password"
        meta={
          Object {
            "error": null,
            "touched": false,
          }
        }
        name="password"
        placeholder="••••••••••"
        type="password"
        value=""
      />
    </FormGroup>
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN IN
    </Button>
  </Form>
  <p
    className="text-center"
  >
    Dont have an account, 
    <Link
      replace={false}
      target="_self"
      to="\\\\signup"
    >
      Sign up here
    </Link>
    .
  </p>
</div>
`;

exports[`LoginForm shows error alert and disables submit button when there's a submit error 1`] = `
<div
  id="signin-form-div"
>
  <h4
    className="text-center"
  >
    Sign In to Your Account
  </h4>
  <hr />
  <Form
    className="mt-4 mb-3 px-5"
    id="signin-form"
    onSubmit={[Function]}
    tag="form"
  >
    <NormalAlert
      color="danger"
    >
      <p
        className="text-center mb-0"
      >
        Username/Password do not match
      </p>
    </NormalAlert>
    <FormGroup
      tag="div"
    >
      <Label
        className="col-form-label"
        for="login-email"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Email Address
      </Label>
      <RenderInput
        handleBlur={[Function]}
        handleChange={[Function]}
        handleFocus={[Function]}
        id="login-email"
        meta={
          Object {
            "asyncValidating": false,
            "error": null,
            "touched": false,
          }
        }
        name="email"
        placeholder="janesmith@me.com"
        type="email"
        value=""
      />
    </FormGroup>
    <FormGroup
      tag="div"
    >
      <Label
        className="col-form-label"
        for="login-password"
        tag="label"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        Password
      </Label>
      <RenderInput
        className="mb-4"
        handleBlur={[Function]}
        handleChange={[Function]}
        handleFocus={[Function]}
        id="login-password"
        meta={
          Object {
            "error": null,
            "touched": false,
          }
        }
        name="password"
        placeholder="••••••••••"
        type="password"
        value=""
      />
    </FormGroup>
    <Button
      className="btn-block mt-0"
      color="secondary"
      disabled={true}
      tag="button"
    >
      SIGN IN
    </Button>
  </Form>
  <p
    className="text-center"
  >
    Dont have an account, 
    <Link
      replace={false}
      target="_self"
      to="\\\\signup"
    >
      Sign up here
    </Link>
    .
  </p>
</div>
`;
